internal/setup/openpgp_test.go

```diff
@@ -0,0 +1,269 @@
+	summary  string
+	armored  string
+	relerror string
+	pubKey   *packet.PublicKey
+}
+
+var archiveKeyTests = []archiveKeyTest{{
+	summary: "Armored data with one public key",
+	armored: testKey.ArmoredPublicKey,
+	pubKey:  testKey.PublicKey,
+}, {
+	summary:  "Armored data with two public keys",
+	armored:  twoPubKeysArmored,
+	relerror: ".*contains more than one public key.*",
+}, {
+	summary:  "Armored data with no public key",
+	armored:  armoredDataWithNoKeys,
+	relerror: ".*no public key.*",
+}, {
+	summary:  "Armored data with private key",
+	armored:  testKey.ArmoredPrivateKey,
+	relerror: ".*contains private key.*",
+}, {
+	summary: "Invalid armored data",
+	armored: `
+		Roses are red
+		Violets are blue
+	`,
+	relerror: ".*cannot decode armored data.*",
+}, {
+	summary:  "Empty armored data",
+	relerror: ".*cannot decode armored data.*",
+}}
+
+func (s *S) TestDecodeArchivePubKey(c *C) {
+	for _, test := range archiveKeyTests {
+		c.Logf("Summary: %s", test.summary)
+
+		pubKey, err := setup.DecodeArchivePublicKey([]byte(test.armored))
+		if test.relerror != "" {
+			c.Assert(err, ErrorMatches, test.relerror)
+			continue
+		} else {
+			c.Assert(err, IsNil)
+		}
+
+		c.Assert(pubKey, DeepEquals, test.pubKey)
+	}
+}
+
+type verifyClearSignTest struct {
+	summary   string
+	clearData string
+	pubKeys   []*packet.PublicKey
+	relerror  string
+}
+
+var verifyClearSignTests = []verifyClearSignTest{{
+	summary:   "Good data with proper sign",
+	clearData: clearSignedData,
+	pubKeys:   []*packet.PublicKey{testKey.PublicKey},
+}, {
+	summary:   "Good data with multiple signatures",
+	clearData: clearSignedWithMultipleSigns,
+	pubKeys:   []*packet.PublicKey{testKey.PublicKey, extraTestKey.PublicKey},
+}, {
+	summary:   "Multiple signatures: verify at least one signature",
+	clearData: clearSignedWithMultipleSigns,
+	pubKeys:   []*packet.PublicKey{testKey.PublicKey},
+}, {
+	summary:   "Invalid data: improper hash",
+	clearData: invalidSignedData,
+	pubKeys:   []*packet.PublicKey{testKey.PublicKey},
+	relerror:  ".*invalid signature: hash tag doesn't match.*",
+}, {
+	summary:   "Invalid data: malformed clearsign text",
+	clearData: "foo\n",
+	pubKeys:   []*packet.PublicKey{testKey.PublicKey},
+	relerror:  ".*invalid clearsign text.*",
+}, {
+	summary:   "Wrong public key to verify with",
+	clearData: clearSignedData,
+	pubKeys:   []*packet.PublicKey{extraTestKey.PublicKey},
+	relerror:  ".*invalid signature:.*verification failure",
+}}
+
+func (s *S) TestVerifySignature(c *C) {
+	for _, test := range verifyClearSignTests {
+		c.Logf("Summary: %s", test.summary)
+
+		sigs, body, _, err := setup.DecodeClearSigned([]byte(test.clearData))
+		if err == nil {
+			// Verify at least one signature with the set of public keys.
+			for _, sig := range sigs {
+				for _, pubKey := range test.pubKeys {
+					err = setup.VerifySignature(pubKey, sig, body)
+					if err == nil {
+						break
+					}
+				}
+				if err == nil {
+					break
+				}
+			}
+		}
```

Comment:
We are reimplementing here the bit of logic that we want to test. That tells me that we should find a way to test the code directly or extract this bit from `archive.go` into its own function in `setup.go` (or any other place) and then test it.

Reply:
Moved this logic to ``setup.VerifyAnySignature``.